.github/workflows/pull_request.yml

```diff
@@ -5,25 +5,64 @@ on:
     types: [opened, synchronize, reopened]
 
 jobs:
-  build:
-    name: Lint & Test
+  docker:
+    name: Build, Lint & Test
     runs-on: ubuntu-latest
     steps:
-      - name: Cancel previous runs
-        uses: styfle/cancel-workflow-action@0.4.0
+      - name: Checkout
+        uses: actions/checkout@v2
+
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v1
+
+      - name: Cache Docker layers
+        uses: actions/cache@v2
         with:
-          access_token: ${{ github.token }}
+          path: /tmp/.buildx-cache
+          key: ${{ runner.os }}-buildx-${{ github.sha }}
+          restore-keys: |
+            ${{ runner.os }}-buildx-
 
-      - name: Clone the repo
-        uses: actions/checkout@v2
+      - name: Login to DockerHub
+        uses: docker/login-action@v1
+        with:
+          username: ${{ secrets.DOCKER_HUB_USERNAME }}
+          password: ${{ secrets.DOCKER_HUB_ACCESS_TOKEN }}
 
-      - name: Build the Docker image
+      - name: Build
+        uses: docker/build-push-action@v2
+        with:
+          context: .
+          push: true
+          tags: codeforafrica/actnow:latest
+          target: python-app-ci
+          cache-from: type=local,src=/tmp/.buildx-cache
+          cache-to: type=local,dest=/tmp/.buildx-cache-new
+
+        # This ugly bit is necessary if you don't want your cache to grow forever
+        # till it hits GitHub's limit of 5GB.
+        # Temp fix
+        # https://github.com/docker/build-push-action/issues/252
+        # https://github.com/moby/buildkit/issues/1896
+      - name: Move cache
+        run: |
+          rm -rf /tmp/.buildx-cache
+          mv /tmp/.buildx-cache-new /tmp/.buildx-cache
+
+      - name: Run app
         run: |
           echo "ACTNOW_SECRET_KEY=${{ secrets.ACTNOW_SECRET_KEY }}" >> .env
-          make run
+          echo "ACTNOW_SENTRY_DSN=${{ secrets.ACTNOW_SENTRY_DSN }}" >> .env
+          make runci
 
-      - name: Test
+      - name: Test app
         run: make test
 
+      - name: Install dev dependencies
+        run: pip3 install -r requirements-dev.txt
```

Comment:
Since we’re targeting `python-app-ci` in the build step above, these should already be installed.

Reply:
I might be wrong but I'm thinking this `run` step is not running inside the docker containers that the `Run app` step starts. The only way we would utilize the already installed packages is if we enter into the running container (Which is what the `test` and the previous `lint` that was causing the OOM were doing) so what we are doing is installing the packages directly into the github server and then telling it to do our linting directly.

Reply:
... that's messed up though. We want to test code in the container. Otherwise, may be we should format, lint, test and _then_ build the image?